### PR TITLE
[DNM] raftstore: add async append

### PR DESCRIPTION
**Pull request by overvenus**
Move **sync** append raft logs out of raftstore, so that raftstore can continue to serve read requests.

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -1087,38 +1139,13 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-            let mut write_opts = WriteOptions::new();
-            write_opts.set_sync(true);
-            self.kv_engine
-                .write_opt(kv_wb, &write_opts)
-                .unwrap_or_else(|e| {
-                    panic!("{} failed to save append state result: {:?}", self.tag, e);
-                });
-        }
-        fail_point!("raft_between_save");
+        let mut new_res = self.poll_append();
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: You can try to handle `ready_res` at L1095.

File: src/raftstore/store/worker/append.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            self.raft_engine
+                .write_opt(task.raft_wb, &write_opts)
+                .unwrap_or_else(|e| {
+                    panic!("{} failed to save raft append result: {:?}", self.tag, e);
+                });
+        }
+        fail_point!("raft_after_save");
+
+        self.notifier
+            .send(TaskRes {
```
Comment (BusyJay) [inline src/raftstore/store/worker/append.rs]: Maybe send a signal to raftstore thread can reduce latency.

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -1055,26 +1090,43 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+                    if !peer.handle_raft_ready_append(&mut ctx, &mut ready_res, &self.pd_worker) {
+                        regions_not_handled.insert(region_id);
+                    }
                 }
             }
-            (ctx.kv_wb, ctx.raft_wb, ctx.ready_res, ctx.sync_log)
-        };
+            self.pending_raft_groups = regions_not_handled;
+
+            // Needs to schedule Ready logs.
```
[the remark this answers is not in the document]
Comment (siddontang) [reply src/raftstore/store/store.rs]: Let's keep the thing simple and use a unified way.

[the remark this answers is not in the document]
Comment (BusyJay) [reply src/raftstore/store/store.rs]: That may hurt performance.

[the remark this answers is not in the document]
Comment (siddontang) [reply src/raftstore/store/store.rs]: We will use sync-log by default later, so it doesn't matter.

And @overvenus tests that the performance is still better for the hybrid read-write case

**Pull request closed without merging**